### PR TITLE
CustomAdapter Annotation

### LINKED ISSUE #115: Make registering custom JsonAdapters easier
It's just a little cumbersome to add custom adapters via SPI. I'm thinking we could add an annotation that we can process to automatically add custom adapters to the generated `GeneratedJsonComponent`.

### PR DESCRIPTION
**Pull request by SentryMan**
Now Custom Adapters can be autoregistered simply by adding an annotation

- creates annotation to register basic custom adapters
- change the adapter base type logic to read the interfaces to retrieve the base type
- add a new method to Types to help check if a generic type can be processed
- change generated factory constant to conform to standard naming conventions

fixes #115

### CODE REVIEW

**Comment by SentryMan** [comment]
hmm wait a sec

**Comment by SentryMan** [comment]
Ok we're good now

File: jsonb-generator/src/test/java/io/avaje/jsonb/generator/models/valid/CustomJsonAdapter.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class CustomJsonAdapter implements JsonAdapter<Example2Packet> {
+
+  public CustomJsonAdapter(Jsonb jsonb) {}
+
+  @Override
+  public void toJson(JsonWriter writer, Example2Packet value) {}
+
+  @Override
+  public Example2Packet fromJson(JsonReader reader) {
+    return null;
```
Comment (rob-bygrave) [inline jsonb-generator/src/test/java/io/avaje/jsonb/generator/models/valid/CustomJsonAdapter.java:20]: I think we should have a real implementation here and exercise it via a test.

Comment (SentryMan) [reply jsonb-generator/src/test/java/io/avaje/jsonb/generator/models/valid/CustomJsonAdapter.java:20]: ok

**Comment by SentryMan** [comment]
ok how about now?

File: jsonb-generator/src/main/java/io/avaje/jsonb/generator/Processor.java [hunk 2]
```diff
@@ -60,13 +66,38 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
...
+
+    for (final var typeElement : ElementFilter.typesIn(elements)) {
+      final var type = typeElement.getQualifiedName().toString();
+      if (CustomAdapterPrism.getInstanceOn(typeElement).isGeneric()) {
+
+        ElementFilter.fieldsIn(typeElement.getEnclosedElements()).stream()
+            .filter(
+                v ->
+                    v.getModifiers().contains(Modifier.STATIC)
+                        && "FACTORY".equals(v.getSimpleName().toString()))
```
Comment (rob-bygrave) [inline jsonb-generator/src/main/java/io/avaje/jsonb/generator/Processor.java:86]: Hmm, breaking change?  equalsIgnoreCase?

Comment (SentryMan) [reply jsonb-generator/src/main/java/io/avaje/jsonb/generator/Processor.java:86]: existing adapters will be unaffected as this is only for custom adapters annotated with the new annotation

**Pull request merged**